### ISSUE TITLE
lifetimebound has a confusing diagnostic about not being a type attribute

### ISSUE THREAD

**Issue by AaronBallman** [llvm/llvm-project#107252]
Given:
```
void func(int * [[clang::lifetimebound]] param);
static_assert((int [[clang::lifetimebound]]) 12);
int * [[clang::lifetimebound]] foo;
```
we issue a "lifetimebound attribute cannot be applied to types" diagnostic for the use in the cast expression, but it can be applied to types as shown by the other cases.

**Comment by MitalAshok** [comment]
Related: #96034 

My understanding is that it's ignored for the other cases as well and Clang doesn't issue the diagnostic like it's supposed to.

**Comment by zygoloid** [comment]
`[[clang::lifetimebound]]` *is* a type attribute, but is supposed to only appertain to function types that are the function declarator in an implicit object member function declaration (that is, it can appear only with the cv-qualifiers and ref-qualifiers for the object parameter, and we consider it to appertain to the object parameter in that case rather than to the function type). Per #96034, some of the appertainment enforcement is missing in the type attribute case.
